compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt

```diff
@@ -268,19 +272,51 @@ internal class ContributedInterfaceSupertypeGenerator(session: FirSession) :
       return emptyList()
     }
 
+    fun removeContribution(classId: ClassId, unmatched: MutableSet<ClassId>) {
+      val removed = contributions.remove(classId)
+      if (removed == null) {
+        unmatched += classId
+      }
+    }
+
+    val typeResolverCache = mutableMapOf<FirClassLikeSymbol<*>, Optional<MetroFirTypeResolver>>()
+
+    fun typeResolverFor(symbol: FirClassLikeSymbol<*>): MetroFirTypeResolver? {
+      return typeResolverCache
+        .getOrPut(symbol) { Optional.ofNullable(typeResolverFactory.create(symbol)) }
+        .getOrNull()
+    }
+
+    // Build a cache of origin class -> contribution classes mappings upfront
+    // This maps from an origin class to all contributions that have @Origin pointing to it
+    // TODO make this lazily computed?
+    val originToContributions = mutableMapOf<ClassId, MutableSet<ClassId>>()
+    for ((parentClassId, _) in contributions) {
+      val parentSymbol = parentClassId.toSymbol(session)?.expectAsOrNull<FirRegularClassSymbol>()
+      if (parentSymbol != null) {
+        val localTypeResolver = typeResolverFor(parentSymbol) ?: continue
+
+        parentSymbol.originClassId(session, localTypeResolver)?.let { originClassId ->
+          originToContributions.getOrPut(originClassId) { mutableSetOf() }.add(parentClassId)
+        }
+      }
+    }
```

Comment:
I can't quite tell, is this recursive for `Origin -> GeneratedAbc -> GeneratedDef`?

Reply:
ahh interesting. I think I'm gonna save crossing that bridge for when someone needs it as that could be expensive in FIR :)